runtime/src/rad_claims.rs

```diff
@@ -0,0 +1,494 @@
+        /// Feeless dispatchable function
+        /// The extrinsic is validated by the custom `validate_unsigned` function below
+        ///
+        /// # <weight>
+        /// - Based on hashes length
+        /// # </weight>
+        #[weight = 195_000_000]
+        pub fn claim(origin,
+                     account_id: T::AccountId,
+                     amount: T::Balance,
+                     sorted_hashes: Vec<T::Hash>,
+        ) -> DispatchResult {
+            ensure_none(origin)?;
+
+            let claimed = Self::get_account_balance(&account_id);
+
+            // Payout = amount - claim
+            let payout = amount.checked_sub(&claimed)
+                .ok_or(Error::<T>::InsufficientBalance)?;
+
+            // Payout must not be less than minimum allowed
+            ensure!(payout >= MIN_PAYOUT.saturated_into(),
+                    Error::<T>::UnderMinPayout);
+
+            let source = MODULE_ID.into_account();
+            // Checking balance of Module account before proceeding
+            let current_balance = T::Currency::free_balance(&source);
+            ensure!(U256::from(current_balance.saturated_into()) >= U256::from(payout.saturated_into()), Error::<T>::InsufficientBalance);
+
+            // Set account balance to amount
+            AccountBalances::<T>::insert(account_id.clone(), amount);
+
+            // Transfer payout amount
+            <pallet_balances::Module<T> as Currency<_>>::transfer(
+                &source,
+                &account_id,
+                payout,
+                AllowDeath,
+            )?;
+
+            Self::deposit_event(RawEvent::Claimed(account_id, amount));
+
+            Ok(())
+        }
+
+        /// Admin function that sets the allowed upload account to add root hashes
+        /// Controlled by custom origin or root
+        /// 
+        /// # <weight>
+        /// - Based on origin check and write op
+        /// # </weight>
+        #[weight = 190_000_000]
+        pub fn set_upload_account(origin, account_id: T::AccountId) -> DispatchResult {
+            Self::can_update_upload_account(origin)?;
+
+            <UploadAccount<T>>::put(account_id);
+
+            Ok(())
+        }
+
+        /// Stores root hash for correspondent claim merkle tree run
+        ///
+        /// # <weight>
+        /// - Based on origin check and write op
+        /// # </weight>
+        #[weight = 185_000_000]
+        pub fn store_root_hash(origin, root_hash: T::Hash) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            ensure!(Self::get_upload_account() == who, Error::<T>::MustBeAdmin);
+            <RootHashes<T>>::insert(root_hash, true);
+            Self::deposit_event(RawEvent::RootHashStored(root_hash));
+
+            Ok(())
+        }
+
+    }
+}
+
+impl<T: Trait> Module<T> {
+    /// Hash a:b if a < b, else b:a. Uses the runtime module's hasher.
+    pub fn sorted_hash_of(a: &T::Hash, b: &T::Hash) -> T::Hash {
```

Comment:
We should repurpose the proofs module for verification.

Reply:
I agree, it would be nice to have a function to generate a document root. Right now, building the tree is coupled with verifying it in the proofs module.

For the `Proofs::sorted_hash_of`, it uses H256, but rad claims merkle tree will be encoded with the runtime hash function. They are both blake2 but H256 can't be elevated to T::Hash associated type. It seems cleaner to just redefine the function rather than change the proofs module and coerce down to H256.